client/src/state/snapshot.rs

```diff
@@ -76,6 +80,101 @@ impl Snapshot {
             Err(_) => f().map_err(|e| e.into()),
         }
     }
+
+    /// Sync two client states.
+    pub fn sync_clients(
+        &mut self,
+        cid0: ClientId,
+        cid1: ClientId,
+        mapper: Option<Mapper<(VaultId, RecordId)>>,
+        merge_policy: SelectOrMerge<SelectOne>,
+    ) {
+        let mut state0 = self.state.0.remove(&cid0).unwrap();
+        let mut state1 = self.state.0.remove(&cid1).unwrap();
+
+        let source: sync::ClientState = (&mut state0).into();
+        let mut target: sync::ClientState = (&mut state1).into();
+
+        let hierarchy = source.get_hierarchy();
+        let diff = target.get_diff(hierarchy, mapper.as_ref(), &merge_policy);
+        let exported = source.export_entries(Some(diff));
+        target.import_entries(exported, &merge_policy, mapper.as_ref(), Some(source.keystore));
+
+        self.state.0.insert(cid0, state0);
+        self.state.0.insert(cid1, state1);
+    }
+
+    /// Sync the local state with another snapshot, which imports the entries from `other` to `local`.
+    pub fn sync_with_snapshot(
+        &mut self,
+        other: &mut SnapshotState,
+        mapper: Option<Mapper<(ClientId, VaultId, RecordId)>>,
+        merge_policy: SelectOrMerge<SelectOrMerge<SelectOne>>,
```

Comment:
If we would only have it on one layer it would be better, I agree. But we have this on three layers:
- Vault layer: only `SelectOne` (select-one between two records)
- Client layer: `SelectOrMerge<SelectOne>` (Select-or-merge between two vaults, select-one between two records)
- Snpashot layer: SelectOrMerge<`SelectOrMerge<SelectOne>>` (Select-or-merge between two clients, Select-or-merge between two vaults, select-one between two records)
If we translate this into one enum each, Imo the recursion in this would be less transparent, and instead we would end up with larger enums (4 variants for client layer, 6 variants on snapshot layer) that are a bit redundant. Furthermore, because the `<SnapshotState as MergeLayer>` implementation internally recursively calls `<ClientState as MergeLayer>`, we would also then have to map between the two types.
That  being said, I don't have a strong opinion on this, so if you insist I can change it to a single enum each.